[blog] flink on zeppelin - part2 (#344)
apache/flink-web: _posts/2020-05-25-flink-on-zeppelin-part2.md

---
layout: post
title:  "Flink on Zeppelin Notebooks for Interactive Data Analysis - Part 2"
date:   2020-05-25T08:00:00.000Z
categories: ecosystem
authors:
- zjffdu:
  name: "Jeff Zhang"
  twitter: "zjffdu"
---

> In the last post, I introduce the basics of Flink on Zeppelin and how to do Streaming ETL. This is part-2 where I would talk about how to 

Reviewer: ```suggestion
In a previous post, we introduced the basics of Flink on Zeppelin and how to do Streaming ETL. In this second part of the "Flink on Zeppelin" series of posts, I will share how to 
```

> do streaming data visualization via Flink on Zeppelin and how to use flink UDF in Zeppelin. 

Reviewer: ```suggestion
perform streaming data visualization via Flink on Zeppelin and how to use Apache Flink UDFs in Zeppelin. 
```

# Streaming Data Visualization

> In Zeppelin, you can build a realtime streaming dashboard without writing any line of javascript/html/css code.

Reviewer: ```suggestion
With [Zeppelin](https://zeppelin.apache.org/), you can build a real time streaming dashboard without writing any line of javascript/html/css code.
```

Overall Zeppelin supports 3 kinds of streaming data analytics:
* Single
* Update

> * Append

Reviewer 2: ```suggestion

Overall, Zeppelin supports 3 kinds of streaming data analytics:

* Single Mode
* Update Mode
* Append Mode
```

### Single Mode

> Single mode is for the case when the result of sql statement is always one row, such as the following example. 

Reviewer: ```suggestion
Single mode is used for cases when the result of a SQL statement is always one row, such as the following example. 
```

> The output format is HTML, and you can specify paragraph local property template for the final output content template. 

Reviewer: ```suggestion
The output format is translated in HTML, and you can specify a paragraph local property template for the final output content template. 
```

> And you can use {i} as placeholder for the ith column of result.

Reviewer: I am not sure what "ith" maybe you want to ad {i}th if you want to say 9th or 8th? 
It should also be "column of **the** result"

<center>
<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_single_mode.gif" width="80%" alt="Single Mode"/>
</center>

### Update Mode

> Update mode is suitable for the case when the output is more than one rows, 

Reviewer: ```suggestion
Update mode is suitable for the cases when the output format is more than one row, 
```

> and always will be updated continuously. Here’s one example where we use group by.

Reviewer: ```suggestion
and will always be continuously updated. Here’s one example where we use group by.
```

> and will always be continuously updated. Here’s one example where we use group by.

Reviewer 2: ```suggestion
and will always be continuously updated. Here’s one example where we use ``GROUP BY``.
```

<center>
<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_update_mode.gif" width="80%" alt="Update Mode"/>
</center>

### Append Mode

> Append mode is suitable for the scenario where output data is always appended. 

Reviewer: ```suggestion
Append mode is suitable for the cases when the output data is always appended. 
```

> E.g. the following example which use tumble window.

Reviewer: ```suggestion
For instance, the example below uses a tumble window.
```

<center>
<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_append_mode.gif" width="80%" alt="Append Mode"/>
</center>

# UDF

> SQL is powerful, especially in expressing data flow. But most of time, you need to handle complicated business logic that can not be expressed by SQL.

Reviewer: ```suggestion
SQL is a very powerful language, especially in expressing data flow. But most of the time, you need to handle complicated business logic that cannot be expressed by SQL.
```

> In these cases, you will need UDF (user defined function). In Zeppelin, you can write Scala, Python UDF, and also import Scala, Python and Java UDF.

Reviewer: ```suggestion
In these cases UDFs (user-defined functions) come particularly handy. In Zeppelin, you can write Scala or Python UDFs, while you can also import Scala, Python and Java UDFs.
```

> Here're 2 examples of Scala and Python UDF.

Reviewer: ```suggestion
Here are 2 examples of Scala and Python UDFs:
```

* Scala UDF

```scala
%flink

class ScalaUpper extends ScalarFunction {
def eval(str: String) = str.toUpperCase
}
btenv.registerFunction("scala_upper", new ScalaUpper())

```
 
* Python UDF

```python

%flink.pyflink

class PythonUpper(ScalarFunction):
def eval(self, s):
 return s.upper()

bt_env.register_function("python_upper", udf(PythonUpper(), DataTypes.STRING(), DataTypes.STRING()))

```

> After you define the UDF, you can use them directly in SQL. e.g.

Reviewer: ```suggestion
After you define the UDFs, you can use them directly in SQL. e.g.
```

> After you define the UDFs, you can use them directly in SQL. e.g.

Reviewer 2: ```suggestion
After you define the UDFs, you can use them directly in SQL:
```

* Use Scala UDF in SQL

<center>
<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_scala_udf.png" width="100%" alt="Scala UDF"/>
</center>

* Use Python UDF in SQL

<center>
<img src="{{ site.baseurl }}/img/blog/2020-05-25-flink-on-zeppelin-part2/flink_python_udf.png" width="100%" alt="Python UDF"/>
</center>

# Summary

> In this post, I show you how to do streaming data visualization via Flink on Zeppelin and how to use UDF. 

Reviewer: ```suggestion
In this post, we explained how to perform streaming data visualization via Flink on Zeppelin and how to use UDFs. 
```

> Besides that, you can do more in Zeppelin with Flink, such as batch processing, hive integration and etc.

Reviewer: ```suggestion
Besides that, you can do more in Zeppelin with Flink, such as batch processing, hive integration and more.
```

> Besides that, you can do more in Zeppelin with Flink, such as batch processing, hive integration and more.

Reviewer 2: ```suggestion
Besides that, you can do more in Zeppelin with Flink, such as batch processing, Hive integration and more.
```

---
layout: post
title:  "Flink on Zeppelin Notebooks for Interactive Data Analysis - Part 2"

> date:   2020-05-25T08:00:00.000Z

Reviewer 2: ```suggestion
date:   2020-06-23T08:00:00.000Z
```

Reviewer 2: @zjffdu, will try to get this one out next week.
